Fix to jer application on jec variations

JulesVandenbroeck opened the pull request:
This pull request intends to fix Jet energy resolution smearing. The issue is that in current implementation the jet energy resolution smearing is applied only to the nominal `Jet.pt` and `Jet.mass` without looking at Jet Energy Correction (JEC) uncertainty variations. The same smearing needs to be also applied to the columns `Jet.{pt,mass}_jec_{jec_uncertainty_name}_{shift}` for each JEC uncertainty variation. In addition the smearing needs to be propagate to MET if `self.propagate_met=True` for each JEC uncertainty variation. This pull request fixes these issues by 're-producing' these columns with the smearing factor applied and the same for MET if requested.

**Checks**
- ~~Currently all JEC uncertainty variations are calculated in the jec producer instead of only the uncertainties in `self.uncertainty_sources`. Should this be changed for efficiency?~~ (edit: I misread the definition of `self.evaluators`)
- Is this implemenation bug-proof?
- ~~Should the possibility for a shift that includes both JEC and JER  uncertainty variations be included?~~

pkausw commented:
Thanks for the PR @JulesVandenbroeck ! I'm tagging @dsavoiu here since he's the one that implemented this. Given that he's a L3 JetMet convener, I'm surprised that this implementation is not correct - do you have a source for this? I know that often times, the resolution smearing and scale corrections for objects are supposed to be applied separately, so it would be good to have a quotable source. Maybe Daniel can also comment on this himself.

JulesVandenbroeck commented:
I don't have a quotable source (yet) as JERC recommendations are discussed separately [here](https://cms-jerc.web.cern.ch/Recommendations/). This came from general intuition of uncertainties that the up/down variation of an uncertainty should still have all other nominal corrections applied. Otherwise you are comparing this up/down variation with effects of other corrrections in there. From a more practical point of view: This means that if an arbitrary correction would shifts the jet pT by 20%  but not for the JEC uncertainty variations on the jet pT of let's say 5%. Both up (+5%) and down (-5%) would have a variatons of 25% and 15% for up and down respectively. This makes this the JEC uncertainty one-sided while it should not be. Below you can find a more subtle example of this happening in my analysis
![image](https://github.com/user-attachments/assets/4dfc4d5a-7279-4eed-8aa2-da49e669af3c)

JulesVandenbroeck commented:
The `smear_factors` are in this pull request calculated with the nominal jet pt even when for the jet pt of each JEC uncertainty variation (up/down). The smear_factor should rather be calculated using the `Jet.pt_jec_{jec_uncertainty_name}_{shift}` column.

riga commented:
I also had a closer look and I think you're right. At the moment, nominal JER is only applied to nominal JEC. Instead, the JER producer should output

- `Jet.{pt,mass}` → central JER evaluated on central JEC
- `Jet.{pt,mass}_jer_{up,down}` → up/down JER evaluated on central JEC
- `Jet.{pt,mass}_jec_SOURCE_{up,down}` → nominal JER evaluated on JEC variations (**currently not done**).

This would not result in additional columns, but just update `Jet.{pt,mass}_jec_SOURCE_{up,down}` columns. Is this what you intended @JulesVandenbroeck ?

Still, we should wait for @dsavoiu 's feedback before taking further action.

dsavoiu commented on columnflow/calibration/cms/jets.py:
```diff
@@ -868,6 +876,22 @@ def jer(self: Calibrator, events: ak.Array, **kwargs) -> ak.Array:
+            # store pt and phi of the full jet system for each jec uncertainty to propagate met
+            if self.propagate_met:
+                jets = events[jet_name]
+                jets = set_ak_column_f32(jets, "pt", jets[f"pt_jec_{name}_{junc_dir}"])
```
Does this create a copy, or does it replace `pt` in the original event array. If the latter then there is a danger that this could produce wrong results. Maybe we could ensure that a copy is made, or better still use a local temporary `dict` to store the shifted `pt` and `mass` columns?

JulesVandenbroeck replied:
Sorry for the late response! I double-checked and `jets = events[jet_name` does create a copy of the original event array. To ensure this is always the case it now takes a deep copy of the original event array.

dsavoiu replied:
Hi @JulesVandenbroeck, thanks for looking into this!

Looks like there may have been some confusion here, which I maybe caused by using the word "copy" a bit too loosely. To clarify, you are correct that the `jets` array returned by the `__getitem__`/`[]` call is a different object than the original array, but it is actually a view of the same underlying data and not a full copy (copying would take too long and make this very inefficient for large arrays). What I meant was whether setting the `pt` field of this object interferes with the original `events` array, as I didn't remember how we implemented it in `set_ak_column`. Luckily, this seems not to be the case, and the `pt` field is just replaced with a different array in the `jets` object, but not in the underlying `events` array.

So it's all fine. And, in fact, adding `ak.copy` here would be counterproductive because it would actually copy the underlying data unnecessarily. Bearing this in mind, let's actually return to the way you had it previously, i.e. without the `ak.copy`. Sorry for the confusion!

dsavoiu commented on columnflow/calibration/cms/jets.py:
```diff
@@ -917,6 +941,30 @@ def jer(self: Calibrator, events: ak.Array, **kwargs) -> ak.Array:
+        for name in self.jec_uncertainty_sources:
+            for junc_dir in ("up", "down"):
+
+                jets = events[jet_name]
+                jets = set_ak_column_f32(jets, "pt", jets[f"pt_jec_{name}_{junc_dir}"])
+                jets = set_ak_column_f32(jets, "mass", jets[f"mass_jec_{name}_{junc_dir}"])
+                jetsum = jets.sum(axis=1)
```
Same as above, it would be better to make a explicit copy here or use local variables.

dsavoiu reviewed (changes requested):
Dear @JulesVandenbroeck, thanks a lot for looking into this and providing a fix! And also to @pkausw and @riga for their input.

You are right of course that the nominal JER smearing should be applied not only to the nominal jet pT but also to the uncertainty-shifted pTs, which seems to have been forgotten when implementing the `jer` producer.

The implementation proposed as part of this PR looks good to me overall, but I do have one concern regarding an apparent in-place array modification which could lead to wrong results (see detailed comments above). Would you be able to run a cross-check for this against an implementation not using the event array for storing the shifted columns @JulesVandenbroeck?

JulesVandenbroeck commented:
@riga @dsavoiu 
Sorry for the delayed response to all your suggestions. I believe I implemented all the suggested changes and also fixed the smearing factor not being recalculated for all the jec variations. If there are remaining suggestions or improvements to be made, let me know!

riga commented:
Highly appreciated @JulesVandenbroeck , thanks!

I'm going to have a look tomorrow.

riga commented:
Hi @JulesVandenbroeck ,

I just had another deeper look and ran the updated recipe for our analysis. Actually, I good with the all the changes (modulo some simplifications that I think are possible), but I noticed that for us (unsurprisingly tbh) the runtime is now seriously increased.

We use deterministic seeds to make the stochastic part random yet reproducible. There are no tools yet to draw from a normal distribution with a vector of seeds in a vectorized fashion, so we have to go for an event loop here which naturally introduces a bottleneck.

With the additional loop here, we do this many times even though we aim to use the same seed per jet. I first thought about moving the random number generation outside your `get_smearing_factor` loop but this somewhat obfuscates the otherwise clean code. A way to avoid this would be to double down on @dsavoiu approach to vectorize over different jet variations (which actually should be removed in the current `get_smearing_factor` since up/down variations of the method itself should not be needed in top of jec uncertainties).

I am currently working on this and I will open a PR to your `upstream/fix_jerc_calibration` branch in a bit, which - when merged - would automatically update *this* PR.

riga reviewed (approved):
This looks good to me now in total.

@dsavoiu With the additional changes, I think a third pair of eyes would be helpful. Would you have the time to give it another pass? (also regarding the two matching strategies and the default). Thank you! :+1:

dsavoiu reviewed (commented):
Thanks a lot @riga for refining the implementation! I think generating one set of RNGs and then using them consistently for smearing all JEC-varied jet pTs is a good approach and avoids adding extra runtime when using deterministic seeds.

I have some small suggestions, mostly regarding code readability, but nothing major. It would be great if you could address them, but otherwise I think this can be merged.

JulesVandenbroeck commented on columnflow/calibration/cms/jets.py:
```diff
@@ -747,294 +744,245 @@ def jer(self: Calibrator, events: ak.Array, **kwargs) -> ak.Array:
     if self.dataset_inst.is_data:
         raise ValueError("attempt to apply jet energy resolution smearing in data")
 
+    # prepare variations
```
This is a good point. Moving `jec_variations`, `jer_variations` and `postfixed` to `jer_init` makes sense. Using these in `uses` and `produces` makes less sence as it is less transparent. 

Since `jer` also produces `"{pt,mass}_unsmeared"` you would need to join the list `postfixes` with `,` instead of a simpleloop: `self.produces.add(f"{self.jet_name}.{{pt,mass}}{{_unsmeared, ','.join(self.postfixes)}}")`. This seems less transparent then what is currently implemented. For `uses` one needs only the nominal and jec variations so `self.postfixes` can not be used

mafrahm replied:
I'd recommend to move them either to the `post_init` or the `setup` function as long as these computations do not change anything in the task dependency tree

JulesVandenbroeck commented on columnflow/calibration/cms/jets.py:
```diff
@@ -747,294 +744,245 @@ def jer(self: Calibrator, events: ak.Array, **kwargs) -> ak.Array:
+    inputs = [variable_map[inp.name] for inp in self.evaluators["jer"].inputs]
+    jerpt = {jer_nom: ak_evaluate(self.evaluators["jer"], *inputs)}
+
+    # for simplifications below, use the same values for jer variations
+    jerpt[jer_up] = jerpt[jer_nom]
+    jerpt[jer_down] = jerpt[jer_nom]
+
```
This is not the case if I am understanding this correctly. `jerpt` contains the `pt` of the jets for each variation (both jer- and jec-variations) that is used in correctionlib. For jer-variations (`{pt,mass}_jer_{up,down}`) this is still the nominal jet pt. The loop over `jec_variations` writes only the jec-variations as keys in `jerpt`, not any variation in `jer_variations`.

riga replied:
I just checked this again :+1:

dsavoiu replied:
Ah right, sorry, I must have misread `jer_up` as `jec_up`. All good, thanks for checking!

JulesVandenbroeck commented on columnflow/calibration/cms/jets.py:
```diff
@@ -747,294 +744,245 @@ def jer(self: Calibrator, events: ak.Array, **kwargs) -> ak.Array:
+    # prepare variations
+    jer_nom, jer_up, jer_down = jer_variations = ["nom", "up", "down"]
+    jec_variations = sum(([f"{unc}_up", f"{unc}_down"] for unc in self.jec_uncertainty_sources), [])
+    assert not (set(jer_variations) & set(jec_variations))
```
Thank you for noticing this. I believe indeed that this assertion is by construction never triggered because. I also included the `jec_` prefix in `jec_variations` in the latest commit so these lists are no longer unique based on just a `_` character (if `unc` is an empty string). Therefore I removed the assertion in this commit

JulesVandenbroeck commented on columnflow/calibration/cms/jets.py:
```diff
@@ -954,13 +951,21 @@ def jer_init(self: Calibrator) -> None:
         self.jec_uncertainty_sources = jec_sources
-        jet_jec_columns = {f"{self.jet_name}.{{pt,mass}}_jec_{jec_source}_{{up,down}}" for jec_source in jec_sources}
-        met_jec_columns = {f"{self.met_name}.{{pt,phi}}_jec_{jec_source}_{{up,down}}" for jec_source in jec_sources}
+
+    # prepare jec variations
+    self.jec_variations = sum(([f"jec_{unc}_up", f"jec_{unc}_down"] for unc in self.jec_uncertainty_sources), [])
+
+    jet_jec_columns = {f"{self.jet_name}.{{pt,mass}}_{jec_source}" for jec_source in self.jec_variations}
+    met_jec_columns = {f"{self.met_name}.{{pt,phi}}_{jec_source}" for jec_source in self.jec_variations}
```
the preparation of jec variations and the sets of columns used in registering the `uses` and `produces` needs to be done even if `jec_sources` is not None.

riga commented on columnflow/calibration/cms/jets.py:
```diff
@@ -747,294 +744,245 @@ def jer(self: Calibrator, events: ak.Array, **kwargs) -> ak.Array:
-        pt_relative_diff = (events[jet_name][pt_name] - matched_gen_jets.pt) / events[jet_name][pt_name]
+    # extract nominal pt resolution
+    inputs = [variable_map[inp.name] for inp in self.evaluators["jer"].inputs]
+    jerpt = {jer_nom: ak_evaluate(self.evaluators["jer"], *inputs)}
```
My bad, I renamed this somewhat too rashly.

riga replied:
I will rename this in the port the v0.3 later today.
